ethereum/eth/src/main/java/tech/pegasys/pantheon/ethereum/eth/sync/tasks/GetReceiptsFromPeerTask.java

```diff
@@ -0,0 +1,126 @@
+ */
+package tech.pegasys.pantheon.ethereum.eth.sync.tasks;
+
+import static java.util.Collections.emptyMap;
+import static java.util.stream.Collectors.toList;
+import static tech.pegasys.pantheon.ethereum.mainnet.BodyValidation.receiptsRoot;
+
+import tech.pegasys.pantheon.ethereum.core.BlockHeader;
+import tech.pegasys.pantheon.ethereum.core.Hash;
+import tech.pegasys.pantheon.ethereum.core.TransactionReceipt;
+import tech.pegasys.pantheon.ethereum.eth.manager.AbstractPeerRequestTask;
+import tech.pegasys.pantheon.ethereum.eth.manager.EthContext;
+import tech.pegasys.pantheon.ethereum.eth.manager.EthPeer;
+import tech.pegasys.pantheon.ethereum.eth.manager.RequestManager.ResponseStream;
+import tech.pegasys.pantheon.ethereum.eth.messages.EthPV63;
+import tech.pegasys.pantheon.ethereum.eth.messages.ReceiptsMessage;
+import tech.pegasys.pantheon.ethereum.p2p.api.MessageData;
+import tech.pegasys.pantheon.ethereum.p2p.api.PeerConnection.PeerNotConnected;
+import tech.pegasys.pantheon.metrics.LabelledMetric;
+import tech.pegasys.pantheon.metrics.OperationTimer;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+public class GetReceiptsFromPeerTask
+    extends AbstractPeerRequestTask<Map<BlockHeader, List<TransactionReceipt>>> {
+  private static final Logger LOG = LogManager.getLogger();
+
+  private final Collection<BlockHeader> blockHeaders;
+  private final Map<Hash, List<BlockHeader>> headersByReceiptsRoot = new HashMap<>();
+
+  private GetReceiptsFromPeerTask(
+      final EthContext ethContext,
+      final Collection<BlockHeader> blockHeaders,
+      final LabelledMetric<OperationTimer> ethTasksTimer) {
+    super(ethContext, EthPV63.GET_RECEIPTS, ethTasksTimer);
+    this.blockHeaders = blockHeaders;
+    blockHeaders.forEach(
+        header ->
+            headersByReceiptsRoot
+                .computeIfAbsent(header.getReceiptsRoot(), key -> new ArrayList<>())
+                .add(header));
+  }
+
+  public static GetReceiptsFromPeerTask forHeaders(
+      final EthContext ethContext,
+      final Collection<BlockHeader> blockHeaders,
+      final LabelledMetric<OperationTimer> ethTasksTimer) {
+
+    return new GetReceiptsFromPeerTask(ethContext, blockHeaders, ethTasksTimer);
+  }
+
+  @Override
+  protected ResponseStream sendRequest(final EthPeer peer) throws PeerNotConnected {
+    LOG.debug("Requesting {} receipts from peer {}.", blockHeaders.size(), peer);
+    // Since we have to match up the data by receipt root, we only need to request receipts
+    // for one of the headers with each unique receipt root.
+    final List<Hash> blockHashes =
+        headersByReceiptsRoot
+            .values()
+            .stream()
+            .map(headers -> headers.get(0).getHash())
+            .collect(toList());
```

Comment:
We probably want to filter out requests for empty receipt sets.  Fine to follow-up with that optimization if you want to go ahead and merge this though.

Reply:
I thought about that but if everything in the list winds up being an empty receipt we wouldn't send a request at all and then we don't have a `ResponseStream` to return and everything gets weird.  So I *think* it's actually better to handle that case back in the calling code - the downside being that it doesn't automatically get re-used if the task winds up called from a second place.

Reply:
hmm - wonder if we could have this return an `Optional<ResponseStream>` or something?  In any case, I say we make a ticket to look into it and just merge this as-is for now.

Reply:
Yeah that's sensible - then just need to have a way for the task to know it should insert the empty result, which probably boils down to just having a flag set or a condition in the process results section.

Reply:
I've added NC-2148 as a subtask so I don't forget this as I work through the actual downloading.